src/Types/OnChainData.cs

```diff
@@ -0,0 +1,144 @@
+    /// <exception cref="InvalidOperationException">The written CBOR data is not accepted under the current conformance mode.</exception>
+    public static OnChainData FromTextEncodeAsCBOR(
+        string data,
+        CborConformanceMode conformanceMode = CborConformanceMode.Strict
+    )
+    {
+        var encoder = new CborWriter(conformanceMode);
+        encoder.WriteTextString(data);
+        var encodedBytes = encoder.Encode();
+        return From(encodedBytes);
+    }
+
+    /// <summary>
+    /// Creates an instance from byte array.
+    /// </summary>
+    /// <param name="data">The data to be registered on-chain represented as a byte array.</param>
+    /// <exception cref="ArgumentException">The length of the supplied data exceeds <see cref="MaxLength"/>.</exception>
+    public static OnChainData From(byte[] dataAsBytes)
+    {
+        if (dataAsBytes.Length > MaxLength)
+        {
+            throw new ArgumentException(
+                $"Size of a data is not allowed to exceed {MaxLength} bytes."
+            );
+        }
+
+        return new OnChainData((byte[])dataAsBytes.Clone());
+    }
+
+    /// <summary>
+    /// Try to decode the data to be registered on-chain as a single CBOR encoded string.
+    /// </summary>
+    /// <returns>
+    /// A <c>string</c> corresponding to the decoded data if it contained
+    /// a single CBOR encoded string, and <c>null</c> otherwise.
+    /// </returns>
+    public string? TryCborDecodeToString()
+    {
+        var encoder = new CborReader(this._value);
+        try
+        {
+            var textRead = encoder.ReadTextString();
+            if (encoder.BytesRemaining == 0)
+            {
+                return textRead;
+            }
+        }
+        catch (CborContentException) { }
+        catch (InvalidOperationException) { }
+        return null;
+    }
+
+    /// <summary>
+    /// Get the on-chain data in the binary format expected by the node.
+    ///
+    /// That is, represented as a byte array with the length of the array
+    /// prepended as a 16-bit unsigned integer in big-endian format.
+    /// </summary>
+    public byte[] GetBytes()
+    {
+        using var memoryStream = new MemoryStream();
+        memoryStream.Write(Serialization.GetBytes((ushort)this._value.Length));
+        memoryStream.Write(this._value);
+        return memoryStream.ToArray();
+    }
+
+    /// <summary>
+    /// Get the data to register on-chain as a hex-encoded string.
+    /// </summary>
+    public override string ToString() => Convert.ToHexString(this._value).ToLowerInvariant();
+
+    public virtual bool Equals(OnChainData? other)
+    {
+        if (other is null)
+        {
+            return false;
+        }
+
+        if (other.GetType() != this.GetType())
+        {
+            return false;
+        }
+
+        return this._value.SequenceEqual(other._value);
+    }
+
+    public override int GetHashCode() => this._value.GetHashCode();
```

Comment:
Just out of curiousity, is the above essentially not as slow as comparison? Not that any of that really matters when the arrays are this small I suppose? Or what?

Reply:
Yes - that’s why in general in .NET you don’t add objects in Hash collections since default is hash on the reference.
 Therefore it is fine to have is like this since normally people don’t work with objects in ways where you would have this performance issue.

If you however overwrite IEquatable you would however expect Equals and GetHash to be aligned.

Reply:
Makes sense. Thanks for the explanation.